File: backend-go/services/memory.go
package services

import "sync"

type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type MemoryStore struct {
	mu    sync.RWMutex
	store map[string][]Message
}

var Memory = &MemoryStore{
	store: make(map[string][]Message),
}

const maxHistory = 20

func (m *MemoryStore) Get(userID string) []Message {
	m.mu.RLock()
	defer m.mu.RUnlock()
	history := m.store[userID]
	result := make([]Message, len(history))
	copy(result, history)
	return result
}

func (m *MemoryStore) Save(userID string, history []Message) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if len(history) > maxHistory {
		history = history[len(history)-maxHistory:]
	}
	m.store[userID] = history
}

func (m *MemoryStore) Clear(userID string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	delete(m.store, userID)
}

File: ai-service/services/memory.py
from threading import Lock

from pydantic import BaseModel


class Message(BaseModel):
    role: str
    content: str


class MemoryStore:
    def __init__(self) -> None:
        self._store: dict[str, list[Message]] = {}
        self._lock = Lock()

    def get(self, user_id: str) -> list[Message]:
        with self._lock:
            return list(self._store.get(user_id, []))

    def save(self, user_id: str, history: list[Message]) -> None:
        with self._lock:
            truncated = history[-20:] if len(history) > 20 else history
            self._store[user_id] = list(truncated)

    def clear(self, user_id: str) -> None:
        with self._lock:
            self._store.pop(user_id, None)


memory = MemoryStore()